storage/backup/backup-cli/src/coordinators/restore.rs

```diff
@@ -88,110 +89,308 @@ impl RestoreCoordinator {
 
-        let next_txn_version = self
+        let target_version = self.global_opt.target_version;
+        COORDINATOR_TARGET_VERSION.set(target_version as i64);
+
+        // calculate the start_version and replay_version
+        let max_txn_ver = metadata_view
+            .max_transaction_version()?
+            .ok_or_else(|| anyhow!("No transaction backup found."))?;
+        let lhs = self.ledger_history_start_version();
+
+        let latest_tree_version = self
+            .global_opt
+            .run_mode
+            .get_state_snapshot_before(Version::MAX);
+        let tree_completed = {
+            match latest_tree_version {
+                Some((ver, _)) => self
+                    .global_opt
+                    .run_mode
+                    .get_state_snapshot_before(ver)
+                    .is_some(),
+                None => false,
+            }
+        };
+
+        let mut db_next_version = self
             .global_opt
             .run_mode
             .get_next_expected_transaction_version()?;
-        if next_txn_version != 0 {
-            // DB is already in workable state
+
+        let kv_snapshot = if db_next_version == 0 {
+            match self.global_opt.run_mode.get_in_progress_state_kv_snapshot() {
+                Ok(Some(ver)) => {
+                    let snapshot = metadata_view.select_state_snapshot(ver)?;
+                    ensure!(
+                        snapshot.is_some() && snapshot.as_ref().unwrap().version == ver,
+                        "cannot find in-progress state snapshot {}",
+                        ver
+                    );
+                    snapshot
+                },
+                Ok(None) | Err(_) => {
+                    metadata_view.select_state_snapshot(std::cmp::min(lhs, max_txn_ver))?
+                },
+            }
+        } else {
+            None
+        };
+
+        let tree_snapshot = if let Some((latest_tree_version, _)) = latest_tree_version {
+            let snapshot = metadata_view.select_state_snapshot(latest_tree_version)?;
+
+            ensure!(
+                snapshot.is_some() && snapshot.as_ref().unwrap().version == latest_tree_version,
+                "cannot find tree snapshot {}",
+                latest_tree_version
+            );
+            snapshot.unwrap()
+        } else {
+            metadata_view
+                .select_state_snapshot(std::cmp::min(self.target_version(), max_txn_ver))?
+                .expect("Cannot find tree snapshot before target version")
+        };
+
+        let two_phase_restore = if let Some(kv_snapshot) = kv_snapshot.as_ref() {
+            // if we have a kv snapshot, we need to restore the state between lhs and rs
+            kv_snapshot.version < tree_snapshot.version
+        } else {
+            // if we don't have a kv snapshot, we need to restore the state between db_next_version and rs
+            db_next_version < tree_snapshot.version && db_next_version > 0
```

Comment:
why it's not "two phase" when db_next_version == 0 and tree_snapshot.version > 0?